src/Estimators/src/InvariantEKFBaseEstimator.cpp

```diff
@@ -0,0 +1,918 @@
+    Eigen::Vector3d sIMU_p_RF = iDynTree::toEigen(IMU_H_RF.getPosition());
+    auto Jl{iDynTree::toEigen(J_IMULF)};
+    auto Jr{iDynTree::toEigen(J_IMURF)};
+
+    Eigen::VectorXd encodersVar = m_sensorsDev.encodersNoise.array().square();
+    Eigen::MatrixXd Renc = static_cast<Eigen::MatrixXd>(encodersVar.asDiagonal());
+
+    // The right invariant observation from the forward kinematics has structure Y = X^{-1} b + V
+    // See Section III C. of the original paper
+    // [ h(s) ]     [R.T -R.T v  -R.T p  -R.T d ] [  0_{3x1}]   [ Jv(s) w_s]
+    // [   0  ]     [         1                 ] [  0      ]   [       0  ]
+    // [   1  ]  =  [                 1         ] [  1      ] + [       0  ]
+    // [  -1  ]     [                          1] [ -1      ]   [       0  ]
+    // s is the encoder measurements and Jv(s) the relevant manipulator Jacobian
+    // Note that the above equation describes only equations for one foot.
+    //
+    // The measurement model Jacobian becomes (shown for one foot),
+    // H = [0_{3x3}  0_{3x3} -I  I]
+    // I is the 3d identity matrix
+    //
+    // The measurement noise covariance,
+    // N = R Jv(s) Cov(w_s) Jv(s).T R.T
+    //
+    // An auxiliary matrix is used along with the Kalman gain due to a reduced form of state update equations
+    // Pi = [I  0_{3x3}]
+    if (meas.lfInContact && meas.rfInContact)
+    {
+        // prepare observation vector Y
+        obs.resize(14);
+        obs << sIMU_p_RF, 0, 1, -1, 0,
+               sIMU_p_LF, 0, 1,  0, -1;
+
+        // prepare measurement model Jacobian H
+        if (m_options.imuBiasEstimationEnabled)
+        {
+            measModelJacobian.resize(6, m_pimpl->m_vecSizeWBias);
+        }
+        else
+        {
+            measModelJacobian.resize(6, m_pimpl->m_vecSizeWOBias);
+        }
+
+        measModelJacobian.topLeftCorner<6, 6>().setZero();
+        measModelJacobian.block<3, 3>(0, m_pimpl->m_vecOffsets.imuPosition) = -Eigen::Matrix3d::Identity();
+        measModelJacobian.block<3, 3>(0, m_pimpl->m_vecOffsets.rContactFramePosition).setIdentity();
+        measModelJacobian.block<3, 3>(0, m_pimpl->m_vecOffsets.lContactFramePosition).setZero();
+        measModelJacobian.block<3, 3>(3, m_pimpl->m_vecOffsets.imuPosition) = -Eigen::Matrix3d::Identity();
+        measModelJacobian.block<3, 3>(3, m_pimpl->m_vecOffsets.rContactFramePosition).setZero();
+        measModelJacobian.block<3, 3>(3, m_pimpl->m_vecOffsets.lContactFramePosition).setIdentity();
+
+        if (m_options.imuBiasEstimationEnabled)
+        {
+            measModelJacobian.block<6, 6>(0, m_pimpl->m_vecOffsets.gyroBias).setZero();
+        }
+
+        // prepare auxiliary gain matrix Pi
+        auxMat.resize(6, 14);
+        auxMat.block<3, 3>(0, 0) = auxMat.block<3, 3>(3, 7) = Eigen::Matrix3d::Identity();
+        auxMat.block<3, 11>(0, 3).setZero();
+        auxMat.block<3, 7>(3, 0).setZero();
+        auxMat.block<3, 4>(3, 10).setZero();
```

Comment:
Super error-prone ;) (by the way it's fine)

Reply:
Error-prone because of the magic numbers? or do you see some other issue? I will remove the magic numbers by proper variable assignments.

Reply:
Magic numbers removed in https://github.com/dic-iit/bipedal-locomotion-framework/pull/85/commits/97ce5cab64ffad935b3f2329824f4cd0d9fc482c